File: frontend/src/types/purchase_order.ts
export type PurchaseOrderStatus = 'draft' | 'approved' | 'sent' | 'partially_received' | 'received' | 'cancelled';

export interface PurchaseOrderItem {
  id: string;
  purchase_order_id: string;
  product_id: string;
  warehouse_id: string | null;
  quantity: number;
  unit_cost: number;
  total_cost: number;
  received_quantity: number;
  line_number: number;
}

export interface PurchaseOrder {
  id: string;
  tenant_id: string;
  order_number: string;
  supplier_id: string;
  status: PurchaseOrderStatus;
  total_amount: number | null;
  currency: string;
  expected_delivery_date: string | null;
  actual_delivery_date: string | null;
  created_by: string;
  created_at: string;
  approved_by: string | null;
  approved_at: string | null;
  sent_at: string | null;
  received_at: string | null;
  cancelled_at: string | null;
  cancelled_by: string | null;
  cancellation_reason: string | null;
  ai_recommendation_id: string | null;
  notes: string | null;
  items: PurchaseOrderItem[];
}

export interface PurchaseOrderItemCreate {
  product_id: string;
  warehouse_id?: string | null;
  quantity: number;
  unit_cost: number;
}

export interface PurchaseOrderCreate {
  supplier_id: string;
  items: PurchaseOrderItemCreate[];
  order_number?: string | null;
  expected_delivery_date?: string | null;
  notes?: string | null;
}

export interface PurchaseOrderReceive {
  received_items: Record<string, number>;  // item_id -> received_quantity
}

File: backend/src/models/purchase_order.py
"""
PurchaseOrder model representing purchase orders.
"""
from sqlalchemy import Column, String, Numeric, Date, DateTime, ForeignKey, UniqueConstraint, Index, CheckConstraint, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy import func
import enum

from src.models.base import BaseModel


class PurchaseOrderStatus(str, enum.Enum):
    """Purchase order status enumeration."""
    DRAFT = "draft"
    APPROVED = "approved"
    SENT = "sent"
    PARTIALLY_RECEIVED = "partially_received"
    RECEIVED = "received"
    CANCELLED = "cancelled"


class PurchaseOrder(BaseModel):
    """PurchaseOrder entity representing purchase orders."""
    
    __tablename__ = "purchase_orders"
    
    order_number = Column(String(100), nullable=False, index=True)
    supplier_id = Column(UUID(as_uuid=True), ForeignKey('suppliers.id'), nullable=False, index=True)
    status = Column(String(20), nullable=False, default=PurchaseOrderStatus.DRAFT.value, index=True)
    total_amount = Column(Numeric(15, 2), nullable=True)
    currency = Column(String(3), nullable=False, default='USD')
    expected_delivery_date = Column(Date, nullable=True)
    actual_delivery_date = Column(Date, nullable=True)
    created_by = Column(UUID(as_uuid=True), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False, index=True)
    approved_by = Column(UUID(as_uuid=True), nullable=True)
    approved_at = Column(DateTime(timezone=True), nullable=True)
    sent_at = Column(DateTime(timezone=True), nullable=True)
    received_at = Column(DateTime(timezone=True), nullable=True)
    cancelled_at = Column(DateTime(timezone=True), nullable=True)
    cancelled_by = Column(UUID(as_uuid=True), nullable=True)
    cancellation_reason = Column(Text, nullable=True)
    ai_recommendation_id = Column(UUID(as_uuid=True), ForeignKey('ai_recommendations.id'), nullable=True, index=True)
    notes = Column(Text, nullable=True)
    
    # Relationships
    supplier = relationship("Supplier", backref="purchase_orders")
    items = relationship("PurchaseOrderItem", back_populates="purchase_order", cascade="all, delete-orphan")
    ai_recommendation = relationship("AIRecommendation", backref="purchase_orders")
    
    # Constraints
    __table_args__ = (
        UniqueConstraint('tenant_id', 'order_number', name='uq_purchase_orders_tenant_order_number'),
        Index('idx_purchase_orders_tenant_order_number', 'tenant_id', 'order_number', unique=True),
        Index('idx_purchase_orders_supplier_id', 'supplier_id'),
        Index('idx_purchase_orders_status', 'status'),
        Index('idx_purchase_orders_created_at', 'created_at'),
        Index('idx_purchase_orders_ai_recommendation_id', 'ai_recommendation_id'),
        CheckConstraint(
            "status IN ('draft', 'approved', 'sent', 'partially_received', 'received', 'cancelled')",
            name='ck_po_status'
        ),
    )
    
    def __repr__(self):
        return f"<PurchaseOrder(id={self.id}, order_number={self.order_number}, status={self.status})>"
